pkg/kf/internal/genericcli/xargs.go

```diff
@@ -0,0 +1,263 @@
+				space := space
+				group.Go(func() error {
+					defer limit.Release(1)
+					return doSpace(groupCtx, cmd, t, listOptions, flags, space, tmpl)
+				})
+			}
+
+			if err := group.Wait(); err != nil {
+				return err
+			}
+
+			if flags.dryRun {
+				fmt.Fprintf(
+					cmd.OutOrStdout(), color.New(color.FgHiYellow).Sprint("Run with --dry-run=false to apply."))
+				fmt.Fprintln(cmd.OutOrStdout())
+			}
+
+			return nil
+		},
+	}
+
+	for flag, label := range options.LabelFilters() {
+		usage := fmt.Sprintf("Sets a filter for the %q label.", label)
+		labelSelectorFlags[label] = cmd.Flags().String(flag, "", usage)
+	}
+
+	flags.Add(cmd)
+
+	return cmd
+}
+
+type XargFlags struct {
+	dryRun           bool
+	allSpaces        bool
+	spaceConcurrency int
+	appConcurrency   int
+}
+
+func (f *XargFlags) Add(cmd *cobra.Command) {
+	cmd.Flags().BoolVar(&f.dryRun, "dry-run", true, "Enables dry-run mode, commands are printed but will not be executed.")
+	cmd.Flags().BoolVar(&f.allSpaces, "all-spaces", false, "Enables targeting all spaces in the cluster.")
+	cmd.Flags().IntVar(&f.spaceConcurrency, "space-concurrency", 1, "Number of spaces that may be operated on in parallel. Only meaningful when multiple spaces are targeted.")
+	cmd.Flags().IntVar(&f.appConcurrency, "app-concurrency", 1, "Number of apps within a space that may be operated on in parallel. Total concurrency will be upto space-concurrency * app-concurrency.")
+}
+
+func doSpace(
+	ctx context.Context,
+	cmd *cobra.Command,
+	t Type,
+	listOptions metav1.ListOptions,
+	flags *XargFlags,
+	space string,
+	commandTemplate *template.Template,
+) error {
+	client := GetResourceInterface(ctx, t, dynamicclient.Get(ctx), space)
+	resource, err := client.List(ctx, listOptions)
+	if err != nil {
+		return err
+	}
+
+	group, ctx := errgroup.WithContext(ctx)
+	limit := semaphore.NewWeighted(int64(flags.appConcurrency))
+
+	for _, item := range resource.Items {
+		limit.Acquire(ctx, 1)
+		if ctx.Err() != nil {
+			break
+		}
+
+		// Shadow item as a local variable to be captured by the closure
+		item := item
+		group.Go(func() error {
+			defer limit.Release(1)
+			b := bytes.NewBuffer([]byte{})
+			err = commandTemplate.Execute(b, &struct {
+				Name  string
+				Space string
+			}{
+				Name:  item.GetName(),
+				Space: item.GetNamespace(),
+			})
+			if err != nil {
+				return err
+			}
+
+			c := b.String()
+
+			w := textio.NewPrefixWriter(cmd.OutOrStdout(), color.New(color.FgGreen).Sprintf("[space=%s app=%s] ", space, item.GetName()))
+			fmt.Fprintf(w, "Exec: %v\n", c)
+
+			if flags.dryRun {
+				return nil
+			}
+
+			p := exec.Command("bash", "-c", c)
+			p.Stdout = w
+			p.Stderr = w
+			if err := p.Start(); err != nil {
+				return err
+			}
+
+			if err := p.Wait(); err != nil {
+				fmt.Fprintf(w, color.New(color.FgHiRed).Sprintf("Command '%v' failed: %v", c, err))
+				fmt.Fprintf(w, color.New(color.Reset).Sprintln())
+				return err
```

Comment:
Can we use errors.Join() to keep processing even in the case of failure and only return a failure at the end?

Reply:
Done, errgroup can be created without cancellation.